src/server/pfs/s3/driver.go

```diff
@@ -52,13 +53,31 @@ func NewMasterDriver() *MasterDriver {
 	return &MasterDriver{}
 }
 
+func branchBucketName(b *pfs.Branch) string {
+	if b.Repo.Type == pfs.UserRepoType {
+		return fmt.Sprintf("%s.%s", b.Name, b.Repo.Name)
+	}
+	return fmt.Sprintf("%s.%s.%s", b.Name, b.Repo.Type, b.Repo.Name)
+}
+
+func projectBranchBucketName(b *pfs.Branch) string {
+	if b.Repo.Type == pfs.UserRepoType {
+		return fmt.Sprintf("%s.%s.%s", b.Name, b.Repo.Name, b.Repo.Project.Name)
+	}
+	return fmt.Sprintf("%s.%s.%s.%s", b.Name, b.Repo.Type, b.Repo.Name, b.Repo.Project.Name)
+}
+
 func (d *MasterDriver) listBuckets(pc *client.APIClient, r *http.Request, buckets *[]*s2.Bucket) error {
+	var isProjectAware = mux.Vars(r)["isProjectAware"] == "yes"
 	repos, err := pc.ListRepoByType("") // get repos of all types
 	if err != nil {
 		return err
 	}
 
 	for _, repo := range repos {
+		if !isProjectAware && repo.Repo.Project.Name != pfs.DefaultProjectName {
```

Comment:
Should this be `isProjectAware` rather than `!isProjectAware`?

Reply:
No: the logic here is that if the client is not project aware and if the repo is not in the default project, then the repo is skipped. So only repos in the default project will be returned to non-project-aware clients.